ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ExternalTableCopyTaskBuilder.java

```diff
@@ -54,14 +59,53 @@
   List<Task<? extends Serializable>> tasks(TaskTracker tracker) {
     List<Task<? extends Serializable>> tasks = new ArrayList<>();
     Iterator<DirCopyWork> itr = work.getPathsToCopyIterator();
-    while (tracker.canAddMoreTasks() && itr.hasNext()) {
+    int numTaskCanBeAdded = tracker.numTaskCanBeAdded();
+    Task<? extends Serializable> barrierTask = TaskFactory.get(new DependencyCollectionWork(), conf);
+    while (numTaskCanBeAdded-- > 0 && itr.hasNext()) {
       DirCopyWork dirCopyWork = itr.next();
       Task<DirCopyWork> task = TaskFactory.get(dirCopyWork, conf);
       tasks.add(task);
-      tracker.addTask(task);
+      barrierTask.addDependentTask(task);
       LOG.debug("added task for {}", dirCopyWork);
     }
-    return tasks;
+
+    if (!tasks.isEmpty()) {
+      tracker.addDependentTask(barrierTask);
+      tracker.addTaskList(tasks);
```

Comment:
why do we need this, all the tasks are already added as dependents to barrier task.

Reply:
removed